### PR TITLE
cl, cmd/capcli: fail the chain download when blobs are not fully stored

### PR DESCRIPTION
**Pull request by lystopad**
Replaces #23623, which targeted code that #23138 has since rewritten. Same defect, correct
target, fresh base.

## The defect

`VerifyAgainstIdentifiersAndInsertIntoTheBlobStore` stops at the first identifier a response
does not match and returns a **nil error** with a partial count. Its only remaining production
caller discarded both counts:

```go
if _, _, err := blob_storage.VerifyAgainstIdentifiersAndInsertIntoTheBlobStore(
    ctx, blobDB, ids, blobs, ...); err != nil {
```

So a block whose blobs stored partially — or not at all — was reported as downloaded. The chain
endpoint moved on to the parent, leaving a gap nothing later notices.

## Change

- check the stored count against the request and fail with both numbers plus the response count
- document the contract on the insert itself, since that count is the only completeness signal
  it gives a caller
- extract `storeRemoteBlobs` so the check is reachable from a test

## Why not #23623

#23623 put the same check in `BlobHistoryDownloader.storeDenebBlobs`. #23138 replaced that whole
path with `denebRecoveryBatch`, which handles it more thoroughly than a warning would: `validate`
returns errors for unrequested, duplicate and empty responses instead of breaking silently,
`store` writes a group only when complete, and `recoverDenebBlobs` re-checks each block with
`storedBlobSidecarsComplete` and calls `addRetrySlot` on a shortfall. The backfill no longer
calls the insert at all, so that half of #23623 is obsolete — and its test targeted a function
that no longer exists.

What #23138 did not change is the insert's contract or the capcli caller, which is what this
picks up.

## Tests

`TestStoreRemoteBlobsRejectsAnIncompleteInsert` drives a sidecar whose header hashes to a
different root, so the insert breaks on the first identifier, stores nothing and returns nil.
`TestStoreRemoteBlobsRejectsAnEmptyResponse` covers the empty-response path to the same end.

Mutation-verified rather than assumed: disabling the count comparison fails both.

`cmd/capcli`, `cl/persistence/blob_storage` and `cl/phase1/network` pass; `make lint` clean.

### CODE REVIEW

**Review by domiwei — CHANGES_REQUESTED** [review]
Thanks for fixing the silent short-insert path. I found two blockers before this is safe to merge:

1. The bootstrap head bypasses blob handling. ChainEndpoint.Run commits the initially fetched /head block and advances currentRoot to its parent before entering loopStep; blob retrieval and storeRemoteBlobs only run inside loopStep. With --blobs, a blob-bearing head can therefore be committed without any sidecars and the command can still return nil. Please route the initial head through the same blob completeness path before committing it, and cover this through ChainEndpoint.Run.

2. The new check uses the count inserted by this invocation rather than durable completeness. Blob storage commits independently before the beacon/index transaction. If the process stops or that transaction fails after all blobs are durable, a restart with a temporary 404/empty response returns inserted=0 and now aborts even though the requested durable set is already complete. Please derive completion from a fresh durable read keyed by the requested identities, and add a pre-populated-store crash/reopen retry test.

The current helper tests pass, but they only assert empty/first-mismatch errors; they do not observe the ChainEndpoint transaction, cursor, durable read, restart, or full-success handoff.

Non-blocking follow-up: malformed JSON sidecars containing null nested headers can still panic before structural validation in VerifyAgainstIdentifiersAndInsertIntoTheBlobStore.

**Review by awskii — APPROVED** [review]
The count guard is right. I checked the cases where it could false-fail and none of them do: `inserted` counts every successful `WriteBlobSidecars`, and writes are unconditional rather than skip-if-present, so re-running the endpoint over already-stored blobs still counts them. A block with no commitments gives `0 < 0`, which is false. A short response leaves the final group unappended, so nothing is written and the guard fires — which is the case you are fixing.

Also confirmed the tests have teeth: removing the `inserted < ids.Len()` check turns both `TestStoreRemoteBlobsRejectsAnIncompleteInsert` and `TestStoreRemoteBlobsRejectsAnEmptyResponse` red.

Documenting the early-stop contract on the insert itself is the right call — that nil error with a partial count is exactly the trap the old caller fell into.

**Review by domiwei — CHANGES_REQUESTED** [review]
Thanks for addressing the bootstrap-head path and the fully indexed restart case. I found three remaining correctness blockers on the current head:

1. `BlobSidecarExists` is only a filesystem `Stat`, so it is not proof that the blob store is durably readable. `WriteBlobSidecars` publishes all sidecar files before it opens and commits the `BlockRootToKzgCommitments` transaction. If cancellation, a DB error, or a crash occurs in that window, a restart plus a 404/empty response makes every check at `cmd/capcli/blob_store.go:88` pass, and the beacon block is committed, while `ReadBlobSidecars` still returns `found=false` because the metadata row is absent. Please make the completion check require the metadata/count row and readable expected sidecars, and add a files-present/metadata-absent restart test through the commit boundary.

2. The expected/requested sets can still diverge. `BlobsIdentifiersFromBlocks` breaks before appending anything when a single decoded block exceeds `MaxRequestBlobSidecarsByVersion`. `ChainEndpoint` decodes but does not consensus-validate the remote block, so a codec-valid Deneb head with 769 commitments produces `ids.Len() == 0`; the insert and existence loops are then empty and the blob-bearing head commits without sidecars. Please require one identifier per commitment (or return an explicit error) and cover the cap+1 case.

3. The extracted path is not Gloas-aware. `GetBlobKzgCommitments` can return non-empty payload-bid commitments for Gloas, but `VerifyAgainstIdentifiersAndInsertIntoTheBlobStore` unconditionally applies the Deneb commitment-inclusion-proof check. The repository already documents the opposite version contract in `TestVerifyBlobSidecarsGloasDoesNotRequireInclusionProof`, so a valid proofless Gloas sidecar is rejected. Please route the block version into version-aware verification and add a focused Gloas success/readback test.

The new focused tests and the full PR CI pass, but they do not exercise these three failure families. The earlier non-blocking nested-null sidecar note also remains; I am not treating that as an additional blocker here.

File: cl/persistence/blob_storage/blob_db.go [hunk 1]
```diff
@@ -288,7 +294,7 @@ func VerifyAgainstIdentifiersAndInsertIntoTheBlobStore(ctx context.Context, stor
 			break
 		}
 
-		if !cltypes.VerifyCommitmentInclusionProof(sidecar.KzgCommitment, sidecar.CommitmentInclusionProof, sidecar.Index, clparams.DenebVersion, sidecar.SignedBlockHeader.Header.BodyRoot) {
+		if version < clparams.GloasVersion && !cltypes.VerifyCommitmentInclusionProof(sidecar.KzgCommitment, sidecar.CommitmentInclusionProof, sidecar.Index, clparams.DenebVersion, sidecar.SignedBlockHeader.Header.BodyRoot) {
```
Comment (awskii) [inline cl/persistence/blob_storage/blob_db.go:297]: The version gate is right, but it arrives here without the check it depends on.

For Gloas the inclusion proof is skipped, and nothing on this path compares `sidecar.KzgCommitment` against `block.Block.Body.GetBlobKzgCommitments().Get(sidecar.Index)`. What remains — header hashes to the identifier root, signature equals `block.Signature`, KZG batch passes — all holds for a self-consistent (blob, commitment, proof) triple that is not the block's. The readback in `storeRemoteBlobs` counts sidecars, so it passes too.

The existing Gloas callers do make that comparison before they reach the gate — `cl/das/peer_das.go:346` rejects on `sidecar.KzgCommitment != metadata.commitments[sidecar.Index]`, and `storedBlobSidecarsComplete` in `cl/phase1/network/blob_downloader.go` does the same. The gate at line 241 is safe *because* of that. This path copies the gate, not the precondition.

Smaller than it sounds in practice: blocks and blobs come from the same `c.Endpoint`, the head is fetched with a nil expected root and never consensus-validated, so an endpoint that lies about blobs could already fabricate a matching block. What is lost is the block↔sidecar consistency check that Deneb and Electra still get — and the asymmetry bites later, since a downstream consumer comparing commitments treats the block as incomplete while capcli has already recorded it as complete.

File: cmd/capcli/blob_store_test.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+	}
+
+	endpoint := &ChainEndpoint{Blobs: true}
+	err := endpoint.storeBlobsForBlock(t.Context(), nil, beaconConfig, "", block)
+
+	require.Error(t, err, "a block whose commitments exceed the request cap must not pass silently")
+	require.ErrorContains(t, err, "identifier")
+}
+
+// Gloas sidecars carry no commitment inclusion proof — VerifyBlobSidecars encodes that contract with
```
Comment (awskii) [inline cmd/capcli/blob_store_test.go]: This pins only the Gloas direction. Dropping the proof check outright, or widening the gate to cover Deneb–Fulu, leaves the suite green: every Deneb call here either passes `nil` sidecars and returns early at `len(sidecars) == 0`, or uses a mismatched header root that breaks the loop before the proof check. Inverting the gate *is* caught, by the zero proof vector.

The neighbouring `TestVerifyBlobSidecarsGloasDoesNotRequireInclusionProof` covers both directions (Gloas NoError, Fulu Error). A pre-Gloas sidecar with a matching root and index but a bad proof, asserted rejected, would give the insert path the same coverage.

File: cmd/capcli/blob_store.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to get blob identifiers: %w", err)
+	}
+	// BlobsIdentifiersFromBlocks drops a whole block once its commitments exceed the per-request
+	// cap, and the remote block is decoded but never consensus-validated, so the endpoint chooses
+	// this count. Without parity the checks below would range over an empty list and pass.
+	if ids.Len() != commitments.Len() {
+		return fmt.Errorf("got %d blob identifiers for %d commitments at slot %d", ids.Len(), commitments.Len(), block.Block.Slot)
+	}
+	blobs, err := retrieveBlobsFromRemoteEndpoint(ctx, beaconConfig, baseUriBlob, block)
```
Comment (awskii) [inline cmd/capcli/blob_store.go:59]: This error no longer names the block. The old `loopStep` wrapped with `fmt.Sprintf("%s/0x%s", baseUriBlob, stringifiedRoot)`; this prints the base `.../eth/v1/beacon/blob_sidecars` path only, and `retrieveBlobsFromRemoteEndpoint` adds neither root nor slot — so a 5xx or a decode failure surfaces as `blob retrieval failed, bad status code 500` with nothing to retry against. The `failed to get blob identifiers` wrap above has no slot either, while the parity and `storeRemoteBlobs` branches below both do. Diagnostics only, but the helper now serves the bootstrap head too, so it is the same gap in two places.

File: cmd/capcli/blob_store.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+	"github.com/erigontech/erigon/cl/clparams"
+	"github.com/erigontech/erigon/cl/cltypes"
+	"github.com/erigontech/erigon/cl/cltypes/solid"
+	"github.com/erigontech/erigon/cl/persistence/blob_storage"
+	"github.com/erigontech/erigon/cl/phase1/network"
+	"github.com/erigontech/erigon/common"
+)
+
+// storeBlobsForBlock fetches and stores the sidecars a block needs, and fails if the store does
```
Comment (awskii) [inline cmd/capcli/blob_store.go:32]: Three of the new comments run into `.claude/rules/comments.md`:

- "Every block the chain download commits goes through here, including the bootstrap head" is a call-site inventory — it says who calls this, not what it guarantees, and goes stale when a commit path moves. The first sentence carries the contract on its own.
- The `TestStoreRemoteBlobsAcceptsAProoflessGloasSidecar` docstring narrates what the old code did. That is commit-message material; the test only needs to say Gloas sidecars carry no inclusion proof.
- "The insert reports a nil error when it stops early" already lives at its canonical place (`blob_db.go:262`). A pointer would do here rather than the full paragraph.

The second paragraph of the `storeRemoteBlobs` docstring should stay — files published before the count row is committed is exactly the non-obvious invariant the rules keep, and it is what stops someone swapping `ReadBlobSidecars` back for a cheaper `Stat`.

**Review by awskii — APPROVED** [review]
All three blockers are genuinely closed. The `Stat` → `ReadBlobSidecars` swap is the substantive one — files are published before the count row commits, so the old check accepted a store nothing could read. Four non-blocking points below; the proof-gate one is the only one I would not leave indefinitely.

File: cmd/capcli/blob_store_test.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+// execution payload bid rather than the body, and NewSignedBeaconBlock leaves the bid unset.
+func commitTo(t *testing.T, block *cltypes.SignedBeaconBlock, commitment common.Bytes48) {
+	t.Helper()
+	list := block.Block.Body.GetBlobKzgCommitments()
+	require.NotNil(t, list, "the block has nowhere to hold commitments")
+	kzgCommitment := cltypes.KZGCommitment(commitment)
+	list.Append(&kzgCommitment)
+}
+
+// Gloas sidecars carry no inclusion proof, so without comparing each sidecar's commitment against
```
Comment (awskii) [inline cmd/capcli/blob_store_test.go:263]: The new bounds check is the only thing standing between an endpoint-supplied index and a panic, and nothing pins it.

`ListSSZ.Get` is `return l.list[index]` (cl/cltypes/solid/list_ssz.go:220) — no bounds guard — and `sidecar.Index` comes straight from the remote JSON. Every sidecar in these tests uses index 0 against one commitment, so the guard is never exercised: I loosened `>=` to `>` and `./cmd/capcli` stayed green.

A response carrying index 1 for a one-commitment block would then crash capcli instead of being rejected. One case with an out-of-range index would hold it.

File: cmd/capcli/blob_store_test.go [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+	blob[31] = blobByte
+	commitment, err := kzg.Ctx().BlobToKZGCommitment(&blob, 0)
+	require.NoError(t, err)
+	proof, err := kzg.Ctx().ComputeBlobKZGProof(&blob, commitment, 0)
+	require.NoError(t, err)
+	return cltypes.NewBlobSidecar(0, (*cltypes.Blob)(&blob), common.Bytes48(commitment),
+		common.Bytes48(proof), header, solid.NewHashVector(cltypes.CommitmentBranchSize))
+}
+
+// commitTo makes block commit to commitment at index 0. For Gloas the commitments live in the
```
Comment (awskii) [inline cmd/capcli/blob_store_test.go:253]: The docstring says `NewSignedBeaconBlock` leaves the Gloas bid unset, but the constructor always sets it — so the sentence describes a state the helper cannot produce.

File: cmd/capcli/blob_store_test.go [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+	require.NoError(t, storeRemoteBlobs(t.Context(), store, ids, []*cltypes.BlobSidecar{sidecar}, block))
+
+	stored, found, err := blob_storage.NewBlobStore(db, fs).ReadBlobSidecars(t.Context(), testBlobSlot, root)
+	require.NoError(t, err)
+	require.True(t, found, "the Gloas sidecar was not durably stored")
+	require.Len(t, stored, 1)
+	require.Equal(t, common.Bytes48(commitment), stored[0].KzgCommitment)
+}
+
+// gloasSidecarFor builds a sidecar with a real KZG commitment and proof but no inclusion proof,
```
Comment (awskii) [inline cmd/capcli/blob_store_test.go:236]: `gloasSidecarFor` repeats the inline setup from the Gloas success test and also builds the pre-Gloas sidecar, so the same construction now lives in two shapes. Worth collapsing to one.

**Review by awskii — APPROVED** [review]
All four points from my last review are closed. The one that mattered is real: `blob_store.go:97` now compares each sidecar's commitment against the block's own, for every version rather than only Gloas, so the insert path finally carries the precondition the version gate always assumed. Three test-quality nits below.

**Review by domiwei — APPROVED** [review]
Reviewed c02d34bc8abd59236da2d04fbfbdb55dda671486. The earlier correctness blockers are resolved: the bootstrap head now crosses the blob-completeness boundary before its beacon transaction commits; completion is based on a consumer-equivalent durable read rather than this invocation's insert count or file existence; identifier/commitment parity prevents a vacuous empty request; and each response sidecar is bounds-checked and bound to the block's commitment before the version-aware proof/KZG path. The focused cmd/capcli and cl/persistence/blob_storage package tests pass, and I found no remaining blocking issue in this PR's scope.
